warehouse/packaging/models.py

```diff
@@ -859,6 +870,31 @@ def __table_args__(cls):  # noqa
     )
     submitted_by: Mapped[User] = orm.relationship(lazy="raise_on_sql")
 
+    @classmethod
+    def create_with_lock(cls, session, *args, **kwargs):
```

Comment:
My issue with this is that now we must remember that the right way to create a `JournalEntry` _must_ use `create_with_lock()`, or we risk creating ids out of order, and there's nothing to prevent someone from doing that, is there?

Had we looked at using the [event listeners we use today on other models](https://github.com/pypi/warehouse/blob/2e4117d127315b42f833939936e6c2e1f84908e7/warehouse/packaging/__init__.py#L38-L59), in this case it would probably be `before_insert`?

If implemented over there, we would take the lock automatically on any insert, thus preserving the normal creation methods we use today.

Reply:
I haven't explored this. Just a note that if we want to go down that route, we'll probably want to do `batch=false` to avoid these being grouped: https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.before_insert:~:text=The%20event%20is%20often%20called%20for%20a%20batch%20of%20objects%20of%20the%20same%20class%20before%20their%20INSERT%20statements%20are%20emitted%20at%20once%20in%20a%20later%20step

Reply:
I switched to a `before_insert` event and verified that the advisory lock query is still being executed prior to the `INSERT`.

I don't believe we need `batch=False`, because we're not actually operating on the individual objects, we just need to insert a single unrelated query _before_ the `INSERT` happens so we can take the advisory lock.